iocore/net/quic/test/test_QUICStreamState.cc

```diff
@@ -206,131 +205,328 @@ TEST_CASE("QUICReceiveStreamState", "[quic]")
     CHECK(ss.is_allowed_to_receive(QUICFrameType::RESET_STREAM));
     ss.update_with_receiving_frame(*rst_stream_frame);
-    CHECK(ss.get() == QUICStreamState::State::ResetRecvd);
+    CHECK(ss.get() == QUICReceiveStreamState::DataRecvd);
   }
 
-  SECTION("Recv -> Size Known -> Reset Recvd -> Data Recvd")
+  SECTION("Recv -> Size Known -> Reset Recvd !-> Data Recvd")
   {
     MockQUICTransferProgressProvider in_progress;
 
     // Case1. Recv STREAM
-    QUICReceiveStreamState ss(&in_progress, nullptr);
+    QUICReceiveStreamStateMachine ss(&in_progress, nullptr);
     CHECK(ss.is_allowed_to_receive(QUICFrameType::STREAM));
     in_progress.set_transfer_progress(1);
     ss.update_with_receiving_frame(*stream_frame);
-    CHECK(ss.get() == QUICStreamState::State::Recv);
+    CHECK(ss.get() == QUICReceiveStreamState::Recv);
 
     // Case2. Recv FIN in a STREAM
     CHECK(ss.is_allowed_to_receive(QUICFrameType::STREAM));
     in_progress.set_transfer_goal(3);
     ss.update_with_receiving_frame(*stream_frame_with_fin);
-    CHECK(ss.get() == QUICStreamState::State::SizeKnown);
+    CHECK(ss.get() == QUICReceiveStreamState::SizeKnown);
 
     // Case3. Recv RESET_STREAM
     CHECK(ss.is_allowed_to_receive(QUICFrameType::RESET_STREAM));
     ss.update_with_receiving_frame(*rst_stream_frame);
-    CHECK(ss.get() == QUICStreamState::State::ResetRecvd);
+    CHECK(ss.get() == QUICReceiveStreamState::ResetRecvd);
 
     // Case4. Recv ALL data
     CHECK(ss.is_allowed_to_receive(QUICFrameType::STREAM));
     in_progress.set_transfer_progress(3);
     ss.update_with_receiving_frame(*stream_frame_delayed);
-    CHECK(ss.get() == QUICStreamState::State::DataRecvd);
+    CHECK(ss.get() == QUICReceiveStreamState::ResetRecvd);
+    CHECK(ss.is_allowed_to_send(QUICFrameType::STOP_SENDING) == false);
+  }
+
+  SECTION("Do not discard STREAM and RESET_STREAM in DataRecvd")
+  {
+    MockQUICTransferProgressProvider in_progress;
+
+    // Case1. Recv STREAM
+    QUICReceiveStreamStateMachine ss(&in_progress, nullptr);
+    CHECK(ss.is_allowed_to_receive(QUICFrameType::STREAM));
+    ss.update_with_receiving_frame(*stream_frame);
+    CHECK(ss.get() == QUICReceiveStreamState::Recv);
+
+    // Case2. Recv FIN in a STREAM
+    CHECK(ss.is_allowed_to_receive(QUICFrameType::STREAM));
+    ss.update_with_receiving_frame(*stream_frame_with_fin);
+    CHECK(ss.get() == QUICReceiveStreamState::SizeKnown);
+
+    // // Case3. Recv ALL data
+    CHECK(ss.is_allowed_to_receive(QUICFrameType::STREAM));
+    in_progress.set_transfer_complete(true);
+    ss.update_with_receiving_frame(*stream_frame_delayed);
+    // ss.update_on_transport_recv_event();
+    CHECK(ss.get() == QUICReceiveStreamState::DataRecvd);
+
+    CHECK(ss.is_allowed_to_receive(QUICFrameType::RESET_STREAM));
+    CHECK(ss.is_allowed_to_receive(QUICFrameType::STREAM));
+    CHECK(ss.is_allowed_to_send(QUICFrameType::STOP_SENDING));
+  }
+}
+
+TEST_CASE("QUICBidState", "[quic]")
```

Comment:
I prefer Bidi (not Bid) as the spec say so.

Reply:
updated!